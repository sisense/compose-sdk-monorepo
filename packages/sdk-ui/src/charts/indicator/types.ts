export type LegacyIndicatorChartTypes = 'numericSimple' | 'numericBar' | 'gauge' | 'ticker';

export type BaseMeasure = {
  size: keyof FontSizes;
  value: number;
  maxWidth?: number;
  maxHeight?: number;
};

export type FitTitleMeasure = {
  width: number;
  string: string;
  key: string;
};

export type FitValueMeasure = {
  width: number;
  string: string;
  key: string;
};

export type FitSecondaryMeasure = {
  width: number;
  titleString: string;
  valueString: string;
  key: string;
};

type FontSizes = {
  big: number;
  medium: number;
  small: number;
  micro: number;
};

type DataKey = 'title' | 'value' | 'secondaryTitle' | 'secondary';

type CommonOptions = {
  fontFamily: string;
  title: {
    color: string;
    fontSizes?: FontSizes;
  };
  value: {
    color?: string;
    fontSizes?: FontSizes;
    fontWeight?: number;
  };
  secondaryTitle: {
    color: string;
    fontSizes?: FontSizes;
  };
  secondaryValue: {
    fontWeight?: number;
    color: string;
    fontSizes?: FontSizes;
  };
  textKeys: string[];
  backgroundColor: string;
  forceTickerView?: boolean;
};

export type NumericSimpleOptions = CommonOptions & {
  borderColor: string;
  borderWidth: number;
  relativeSizes: {
    key: string;
    value: number;
    decimals: number;
  }[];
  measureKeys: string[];
  sizes: {
    size: keyof FontSizes;
    value: number;
  }[];
  fitMeasures: {
    key: string;
    prop: string;
    dataKey: DataKey;
  }[];
};

export type NumericBarOptions = CommonOptions & {
  bracketColor: string;
  bracketThickness: number;
  relativeSizes: {
    key: string;
    value: number;
    decimals: number;
  }[];
  measureKeys: string[];
  sizes: {
    size: keyof FontSizes;
    value: number;
    maxWidth?: number;
    maxHeight?: number;
  }[];
};

export type GaugeOptions = CommonOptions & {
  label: {
    color: string;
    fontSizes: FontSizes;
  };
  textKeys: string[];
  tickColor: string;
  needleColor: string;
  bracketColor: string;
  defaultDialColor: string;
  gaugeOpacity: number;
  startAngle: number;
  endAngle: number;
  overDegrees: number;
  tickDegreesIncrement: number;
  bracketThickness: number;
  relativeSizes: (
    | {
        key: string;
        value: number;
        decimals: number;
      }
    | {
        key: string;
        dataKey: 'skin';
        values: Record<'1' | '2', number>;
        decimals: number;
      }
  )[];
  gaugeHeights: BaseMeasure[];
  measureKeys: string[];
  tickerBarHeight?: number;
};

export type TickerOptions = CommonOptions & {
  fontSize: number;
  dividerColor: string;
  barHandleColor: string;
  height: number;
  horizontalMargin: number;
  barWidth: number;
  barHeight: number;
  barOpacity: number;
  tickerBarWidth: number;
  tickerBarHeight: number;
  horizontalPadding: number;
  textPadding: number;
  sectionMinWidth: number;
  dividerWidth: number;
  dividerHeight: number;
  valueFont?: string;
  titleFont?: string;
  secondaryValueFont?: string;
  secondaryTitleFont?: string;
  offsetX?: number;
  showTickerBar?: boolean;
  fitValueMeasure?: {
    string: string;
    width: number;
  };
  fitSecValueMeasure?: {
    string: string;
    width: number;
  };
  fitSecTitleMeasure?: {
    string: string;
    width: number;
  };
  fitTitleMeasure?: {
    string: string;
    width: number;
  };
  width?: number;
};

export type LegacyIndicatorChartOptions =
  | NumericBarOptions
  | NumericSimpleOptions
  | GaugeOptions
  | TickerOptions;
