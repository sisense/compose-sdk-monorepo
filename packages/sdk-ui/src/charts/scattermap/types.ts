export type Coordinates = {
  lat: number;
  lng: number;
};

export type Location = {
  _id: string;
  lookupKey: string;
  placetype: null | string;
  context: null | string;
  err: null | string;
  latLng: Coordinates;
  name: string;
  place_name: string;
  text: string;
  version: string;
};
