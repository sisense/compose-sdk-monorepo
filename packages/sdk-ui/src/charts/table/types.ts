import { TableDataOptionsInternal } from '../../chart-data-options/types';
import { Column, DataTable } from '../../chart-data-processor/table-processor';
import {
  TableColorOptions,
  TableDesignOptions,
} from '../../chart-options-processor/translations/design-options';
import { CompleteThemeSettings } from '../../types';

/**
 * Properties needed to setup Table
 *
 * @internal
 */
export type TableProps = {
  /**
   * Explicit Data, which is made up of an array of columns and a two-dimensional array of data cells.
   *
   */
  dataTable: DataTable;
  /**
   * Configurations for how to interpret and present the data passed to the chart
   *
   * @category Chart
   */
  dataOptions: TableDataOptionsInternal;
  /**
   * Configuration that defines functional style of the various chart elements
   *
   */
  designOptions?: TableDesignOptions;
  themeSettings: CompleteThemeSettings;
  /**
   * Callback that calls when sort of column triggered
   *
   */
  onSortUpdate: (column: Column) => void;
  /**
   * Define width of table
   *
   */
  width?: number;
  /**
   * Define height of table
   *
   */
  height?: number;
};

export type DataTableWrapperProps = {
  dataTable: DataTable;
  dataOptions: TableDataOptionsInternal;
  height: number;
  width: number;
  customStyles?: TableCustomStyles;
  themeSettings: CompleteThemeSettings;
  onSortUpdate: (column: Column) => void;
};

export type TableCustomStyles = {
  showFieldTypeIcon?: boolean;
  sortIcon?: 'caret' | 'standard';
  headerHeight?: number;
  rowHeight?: number;
  paddingVertical?: number;
  paddingHorizontal?: number;
  header?: {
    color?: TableColorOptions;
  };
  columns?: {
    alternatingColor?: TableColorOptions;
    width?: 'auto' | 'content';
  };
  rows?: {
    alternatingColor?: TableColorOptions;
  };
};

export type TableReducerState = {
  sortedDataTable: DataTable;
  sortedColumn: Column | undefined;
  sortColumns: Column[];
};

export type TableReducerBaseAction = { type: 'sortBy'; column: Column };

export type TableReducerResetAction = {
  type: 'reset';
  dataTable: DataTable;
  defaultSortColumn?: Column;
};

export type TableReducerAction = TableReducerBaseAction | TableReducerResetAction;

export type TableReducerStateWithItems = TableReducerState & { items: readonly any[] };

export type TableReducerActionWithItems =
  | TableReducerBaseAction
  | (TableReducerResetAction & { items: readonly any[] });
