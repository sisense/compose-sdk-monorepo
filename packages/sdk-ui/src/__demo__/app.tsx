import { Alert, Tab, Tabs } from '@mui/material';
import { Box } from '@mui/system';
import { useState } from 'react';
import { SisenseContextProvider } from '../components/SisenseContextProvider';
import { SisenseContextProviderProps } from '../props';
import { ChartsFromExampleApp } from './pages/ChartsFromExampleApp';
import { ECommerceDemo } from './pages/ECommerceDemo';
import { MiscDemo } from './pages/MiscDemo';
import { WidgetDemo } from './pages/WidgetDemo';

// This page is meant to enable faster iterations during development than
// using react-ts-demo or other demo apps that require a built sdk-ui
// Simply add a page to the `pages` array and it will show up in `yarn dev`
// Suggest adding a router or at least sessionStorage var for selectedTabIndex
// if this becomes popular
const pages = [WidgetDemo, ECommerceDemo, ChartsFromExampleApp, MiscDemo];

const { VITE_APP_SISENSE_URL, VITE_APP_SISENSE_USERNAME, VITE_APP_SISENSE_PASSWORD } = import.meta
  .env;

export function App() {
  const [selectedTabIndex, setSelectedTabIndex] = useState(0);

  const shouldShowAlert =
    !VITE_APP_SISENSE_URL || !VITE_APP_SISENSE_USERNAME || !VITE_APP_SISENSE_PASSWORD;

  const sisenseContextProps: SisenseContextProviderProps = {
    url: VITE_APP_SISENSE_URL ?? '',
    username: VITE_APP_SISENSE_USERNAME,
    password: VITE_APP_SISENSE_PASSWORD,
    defaultDataSource: 'Sample ECommerce',
  };

  return (
    <>
      {shouldShowAlert && (
        <Alert severity="warning">
          One or more expected env vars are missing. You may want to check your
          <code>.env.local</code> file.
        </Alert>
      )}
      <SisenseContextProvider {...sisenseContextProps}>
        <div style={{ width: '100vw', height: '100vh' }}>
          <Box sx={{ borderBottom: 1, borderColor: 'divider' }}>
            <Tabs
              value={selectedTabIndex}
              onChange={(e, value: number) => setSelectedTabIndex(value)}
            >
              {pages.map((page, i) => (
                <Tab key={i} label={page.name} sx={{ textTransform: 'none' }} />
              ))}
            </Tabs>
          </Box>
          {pages.map((Page, i) => (
            <div key={i} hidden={selectedTabIndex !== i}>
              {selectedTabIndex === i && (
                <Box sx={{ p: 3 }}>
                  <Page />
                </Box>
              )}
            </div>
          ))}
        </div>
      </SisenseContextProvider>
    </>
  );
}
