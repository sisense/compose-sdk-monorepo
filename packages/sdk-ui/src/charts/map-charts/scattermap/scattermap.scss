.csdk-scattermap-tooltip-container {
  color: #5b6372;
  margin: 4px 6px;
}

.csdk-scattermap-tooltip-content {
  text-align: left;
}

.csdk-scattermap-tooltip-label {
  font-size: 15px;
  line-height: 18px;
}

.csdk-scattermap-tooltip-category {
  color: #9ea2ab;
  padding-right: 20px;
}

.csdk-scattermap-tooltip-row {
  display: flex;
  justify-content: space-between;
}
