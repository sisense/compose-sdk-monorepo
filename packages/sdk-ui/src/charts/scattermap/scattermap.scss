.csdk-scattermap-tooltip-row {
  display: flex;
  justify-content: space-between;
}

.csdk-scattermap-root {
  .leaflet-interactive {
    outline: none;
  }
}
