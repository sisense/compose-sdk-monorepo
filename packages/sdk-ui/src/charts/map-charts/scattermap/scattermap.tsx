/* eslint-disable max-lines-per-function */
import 'leaflet/dist/leaflet.css';
import leaflet from 'leaflet';
import { useEffect, useMemo, useRef } from 'react';
import { useLocations } from './hooks/use-locations.js';
import { ScattermapChartData } from '../../../chart-data/types.js';
import { getLocationsMarkerSizes } from './utils/size.js';
import { useGeoSettings } from './hooks/use-settings.js';
import { getLocationsMarkerColors } from './utils/color.js';
import { createMarker, removeMarkers } from './utils/markers.js';
import { addCopyright } from './utils/copyright.js';
import { fitMapToBounds } from './utils/map.js';
import { ScattermapChartDataOptionsInternal } from '../../../chart-data-options/types.js';
import { ScattermapChartDesignOptions } from '../../../chart-options-processor/translations/design-options.js';
import { useTooltipHandler } from './hooks/use-tooltip-handler.js';
import { DataSource, Filter, FilterRelation, getFilterListAndRelations } from '@sisense/sdk-data';

import '../map-charts.scss';
import './scattermap.scss';

export type ScattermapProps = {
  chartData: ScattermapChartData;
  dataOptions: ScattermapChartDataOptionsInternal;
  designOptions: ScattermapChartDesignOptions;
  dataSource: DataSource | null;
  filters?: Filter[] | FilterRelation;
};

export const Scattermap = ({
  chartData,
  dataOptions,
  dataSource,
  filters = [],
  designOptions,
}: ScattermapProps) => {
  const { locations } = chartData;
  const geoSettings = useGeoSettings();

  const locationsWithCoordinates = useLocations(locations, dataOptions.locationLevel);

  const { filters: filterList } = getFilterListAndRelations(filters);
  const tooltipHandler = useTooltipHandler({
    dataOptions,
    dataSource,
    filters: filterList,
  });

  const mapContainer = useRef(null);
  const mapInstance = useRef<leaflet.Map | null>(null);

  const markerSizes = useMemo(
    () =>
      getLocationsMarkerSizes(
        locationsWithCoordinates,
        dataOptions.size
          ? designOptions.markers.size.minSize
          : designOptions.markers.size.defaultSize,
        dataOptions.size
          ? designOptions.markers.size.maxSize
          : designOptions.markers.size.defaultSize,
      ),
    [locationsWithCoordinates, dataOptions, designOptions],
  );
  const markerColors = useMemo(
    () => getLocationsMarkerColors(locationsWithCoordinates, dataOptions.colorBy?.color),
    [locationsWithCoordinates, dataOptions],
  );

  const markersRef = useRef<leaflet.CircleMarker[]>([]);
  const markersLayerRef = useRef(leaflet.layerGroup([]));

  const tileLayer = useRef<leaflet.TileLayer | null>(null);
  const copyrightRef = useRef<leaflet.Control.Attribution | null>(null);

  useEffect(() => {
    if (mapContainer.current && !mapInstance.current) {
      mapInstance.current = leaflet.map(mapContainer.current, {
        attributionControl: false,
        scrollWheelZoom: true,
        minZoom: 1,
        worldCopyJump: true,
      });

      markersLayerRef.current.addTo(mapInstance.current);
      mapInstance.current.fitWorld();
    }
  }, []);

  useEffect(() => {
    if (markersRef.current) {
      removeMarkers(markersRef.current);
      markersRef.current = [];
    }

    if (locationsWithCoordinates) {
      locationsWithCoordinates.forEach((locationWithCoordinates, index) => {
        if (!locationWithCoordinates.coordinates) return;

        const marker = createMarker({
          coordinates: locationWithCoordinates.coordinates,
          style: {
            color: markerColors[index],
            size: markerSizes[index],
            fill: designOptions.markers.fill,
            blur: locationWithCoordinates.blur,
          },
        });

        markersRef.current.push(marker);
        marker.addTo(markersLayerRef.current);

        marker.bindTooltip(() => {
          const { content, postponedContent } = tooltipHandler(locationWithCoordinates);

          postponedContent?.then((updatedContent) => marker.setTooltipContent(updatedContent));

          return content;
        });
      });

      fitMapToBounds(mapInstance.current!, markersRef.current);
    }
  }, [locationsWithCoordinates, markerColors, markerSizes, designOptions, tooltipHandler]);

  useEffect(() => {
    if (!geoSettings) {
      return;
    }
    if (tileLayer.current) {
      tileLayer.current.remove();
    }

    tileLayer.current = leaflet.tileLayer(geoSettings.maps_api_provider, {
      zoomOffset: -1,
      tileSize: 512,
    });

    if (mapInstance.current) {
      tileLayer.current.addTo(mapInstance.current);

      copyrightRef.current?.remove();
      copyrightRef.current = addCopyright(mapInstance.current, geoSettings.maps_api_provider);
    }
  }, [geoSettings]);

  return (
    <div
      className="csdk-map-container"
      ref={mapContainer}
      style={{
        width: '100%',
        height: '100%',
      }}
    />
  );
};
