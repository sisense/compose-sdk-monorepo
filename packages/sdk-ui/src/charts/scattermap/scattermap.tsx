/* eslint-disable max-lines-per-function */
import 'leaflet/dist/leaflet.css';
import leaflet from 'leaflet';
import { useEffect, useMemo, useRef } from 'react';
import { useLocations } from './hooks/use-locations.js';
import { ScattermapChartData } from '../../chart-data/types.js';
import { getLocationsMarkerSizes } from './utils/size.js';
import { useGeoSettings } from './hooks/use-settings.js';
import { getLocationsMarkerColors } from './utils/color.js';
import { createMarker, removeMarkers } from './utils/markers.js';
import { addCopyright } from './utils/copyright.js';
import { fitMapToBounds } from './utils/map.js';
import { ScattermapChartDataOptionsInternal } from '../../chart-data-options/types.js';
import { ScattermapChartDesignOptions } from '../../chart-options-processor/translations/design-options.js';
import { useTooltipHandler } from './hooks/use-tooltip-handler.js';
import { DataSource, Filter, FilterRelation, getFilterListAndRelations } from '@sisense/sdk-data';

import './scattermap.scss';

export type ScattermapProps = {
  chartData: ScattermapChartData;
  dataOptions: ScattermapChartDataOptionsInternal;
  designOptions: ScattermapChartDesignOptions;
  dataSource: DataSource | null;
  filters?: Filter[] | FilterRelation;
};

export const Scattermap = ({
  chartData,
  dataOptions,
  dataSource,
  filters = [],
  designOptions,
}: ScattermapProps) => {
  const { locations } = chartData;
  const geoSettings = useGeoSettings();

  const locationsProps = useMemo(() => locations.map((l) => l.name), [locations]);
  const locationsWithCoordinates = useLocations(locationsProps, dataOptions.locationLevel);

  const { filters: filterList } = getFilterListAndRelations(filters);

  const tooltipHandler = useTooltipHandler({
    dataOptions,
    dataSource,
    filters: filterList,
  });

  const mapElement = useRef(null);
  const mapInstance = useRef<leaflet.Map | null>(null);

  const markerSizes = useMemo(
    () =>
      getLocationsMarkerSizes(
        locations,
        dataOptions.size
          ? designOptions.markers.size.minSize
          : designOptions.markers.size.defaultSize,
        dataOptions.size
          ? designOptions.markers.size.maxSize
          : designOptions.markers.size.defaultSize,
      ),
    [locations, dataOptions, designOptions],
  );
  const markerColors = useMemo(
    () => getLocationsMarkerColors(locations, dataOptions.colorBy?.color),
    [locations, dataOptions],
  );

  const markersRef = useRef<leaflet.CircleMarker[]>([]);
  const markersLayerRef = useRef(leaflet.layerGroup([]));

  const tileLayer = useRef<leaflet.TileLayer | null>(null);
  const copyrightRef = useRef<leaflet.Control.Attribution | null>(null);

  useEffect(() => {
    if (mapElement.current && !mapInstance.current) {
      mapInstance.current = leaflet.map(mapElement.current, {
        attributionControl: false,
        scrollWheelZoom: true,
        minZoom: 1,
        worldCopyJump: true,
      });

      markersLayerRef.current.addTo(mapInstance.current);
      mapInstance.current.fitWorld();
    }
  }, []);

  useEffect(() => {
    if (markersRef.current) {
      removeMarkers(markersRef.current);
      markersRef.current = [];
    }

    if (locationsWithCoordinates) {
      locationsWithCoordinates.forEach((locationWithCoordinates, index) => {
        if (!locationWithCoordinates) return;

        const marker = createMarker({
          coordinates: locationWithCoordinates.latLng,
          style: {
            color: markerColors[index],
            size: markerSizes[index],
            fill: designOptions.markers.fill,
            blur: locations[index].blur,
          },
        });

        markersRef.current.push(marker);
        marker.addTo(markersLayerRef.current);

        marker.bindTooltip(() => {
          const { content, postponedContent } = tooltipHandler(locations[index]);

          postponedContent?.then((updatedContent) => marker.setTooltipContent(updatedContent));

          return content;
        });
      });

      fitMapToBounds(mapInstance.current!, markersRef.current);
    }
  }, [
    locationsWithCoordinates,
    locations,
    markerColors,
    markerSizes,
    designOptions,
    tooltipHandler,
  ]);

  useEffect(() => {
    if (!geoSettings) {
      return;
    }
    if (tileLayer.current) {
      tileLayer.current.remove();
    }

    tileLayer.current = leaflet.tileLayer(geoSettings.maps_api_provider, {
      zoomOffset: -1,
      tileSize: 512,
    });

    if (mapInstance.current) {
      tileLayer.current.addTo(mapInstance.current);

      copyrightRef.current?.remove();
      copyrightRef.current = addCopyright(mapInstance.current, geoSettings.maps_api_provider);
    }
  }, [geoSettings]);

  return (
    <div
      className="csdk-scattermap-root"
      ref={mapElement}
      style={{
        width: '100%',
        height: '100%',
      }}
    />
  );
};
