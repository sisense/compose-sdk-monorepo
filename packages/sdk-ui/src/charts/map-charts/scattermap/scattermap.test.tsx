import { fireEvent, render, waitFor } from '@testing-library/react';
import { Scattermap } from '@/charts/map-charts/scattermap/scattermap';
import { ScattermapChartDataOptionsInternal } from '@/chart-data-options/types';
import { scattermapData } from '@/chart-data/scattermap-data';
import { ScattermapChartDesignOptions } from '@/chart-options-processor/translations/design-options';
import { DataTable } from '@/chart-data-processor/table-processor';
import { SisenseContext } from '@/sisense-context/sisense-context';
import { HttpClient } from '@sisense/sdk-rest-client/src/http-client';
import { ClientApplication } from '@/app/client-application';

const dataSet = {
  columns: [
    { name: 'Country', type: 'string', index: 0 },
    { name: 'Lat', type: 'number', index: 1 },
    { name: 'Lng', type: 'number', index: 2 },
    { name: 'Quantity', type: 'number', index: 3 },
  ],
  rows: [
    [
      { displayValue: 'Canada' },
      { displayValue: 55.5859012851966 },
      { displayValue: -105.750595856519 },
      { displayValue: 6781 },
    ],
    [
      { displayValue: 'United States' },
      { displayValue: 37.201902 },
      { displayValue: -113.187854 },
      { displayValue: 4012 },
    ],
    [
      { displayValue: 'Albania' },
      { displayValue: 40.641089555859 },
      { displayValue: 20.1566908111252 },
      { displayValue: 4471 },
    ],
    [
      { displayValue: 'Algeria' },
      { displayValue: 28.1632395923063 },
      { displayValue: 2.63238813336793 },
      { displayValue: 1812 },
    ],
    [
      { displayValue: 'Argentina' },
      { displayValue: -36.252002 },
      { displayValue: -63.954193 },
      { displayValue: 5001 },
    ],
    [
      { displayValue: 'Bermuda' },
      { displayValue: 32.3179203152518 },
      { displayValue: -64.7370103996652 },
      { displayValue: 4001 },
    ],
    [
      { displayValue: 'Kyrgyzstan' },
      { displayValue: 41.465053955426 },
      { displayValue: 74.5555962804371 },
      { displayValue: 4242 },
    ],
    [
      { displayValue: 'Ukraine' },
      { displayValue: 49.3227937844972 },
      { displayValue: 31.3202829593814 },
      { displayValue: 936 },
    ],
  ],
} as unknown as DataTable;

const cat1 = {
  column: {
    name: 'Country',
    type: 'string',
  },
};
const cat2 = {
  column: { name: 'Lat', type: 'number' },
};
const cat3 = {
  column: {
    name: 'Lng',
    type: 'number',
  },
};

const meas1 = { column: { name: 'Quantity', aggregation: 'sum' } };

const designOptions = {
  markers: {
    size: {
      minSize: 5,
      maxSize: 15,
      defaultSize: 10,
    },
    fill: 'filled',
  },
} as ScattermapChartDesignOptions;
describe('Scattermap Chart', () => {
  it('render a scattermap with coordinates', async () => {
    const dataOptions = {
      locations: [cat2, cat3],
      size: meas1,
      locationLevel: 'country',
      colorBy: meas1,
    } as ScattermapChartDataOptionsInternal;
    render(
      <Scattermap
        dataSource={'test'}
        dataOptions={dataOptions}
        chartData={scattermapData(dataOptions, dataSet)}
        designOptions={designOptions}
      />,
    );

    await waitFor(() => {
      const map = document.querySelector('.csdk-map-container');
      expect(map).toBeDefined();
      const markers = document.querySelectorAll('.leaflet-interactive');
      expect(markers.length).toBe(8);

      Array.from(markers).forEach((marker) => expect(marker.getAttribute('fill')).toBe('#00cee6'));
    });
  });

  it('render a scattermap with geo names', async () => {
    const dataOptions = {
      locations: [cat1],
      size: meas1,
      locationLevel: 'auto',
    } as ScattermapChartDataOptionsInternal;
    const chartData = scattermapData(dataOptions, dataSet);

    const { findByText } = render(
      <SisenseContext.Provider
        value={{
          app: {
            httpClient: {
              post: () => {
                return Promise.resolve(
                  chartData.locations
                    // Filter one of values to pass missing item response case
                    .filter((location) => location.name !== 'Kyrgyzstan')
                    .map((location) => {
                      return {
                        _id: '1',
                        lookupKey: '',
                        placetype: null,
                        context: null,
                        err: null,
                        latLng: {
                          lat: 1,
                          lng: 1,
                        },
                        name: location.name,
                        place_name: location.name,
                        text: location.name,
                        version: '1',
                      };
                    }),
                );
              },
              get: () => {
                return Promise.resolve({
                  geo: {
                    maps_api_provider: 'https://api.mapbox.com/',
                  },
                });
              },
            } as unknown as HttpClient,
          } as ClientApplication,
          isInitialized: true,
          tracking: {
            enabled: false,
            packageName: 'sdk-ui',
          },
          errorBoundary: {
            showErrorBox: true,
          },
        }}
      >
        <Scattermap
          dataSource={'test'}
          dataOptions={dataOptions}
          chartData={chartData}
          designOptions={designOptions}
        />
      </SisenseContext.Provider>,
    );

    await waitFor(async () => {
      const map = document.querySelector('.csdk-map-container');
      expect(map).toBeDefined();

      const copyright = await findByText('OpenStreetMap');
      expect(copyright).toBeDefined();

      const markers = document.querySelectorAll('.leaflet-interactive');
      expect(markers.length).toBe(7);
    });
  });

  it('render a scattermap and call tooltip', async () => {
    const dataOptions = {
      locations: [cat2, cat3],
      size: meas1,
      locationLevel: 'country',
    } as ScattermapChartDataOptionsInternal;
    render(
      <Scattermap
        dataSource={'test'}
        dataOptions={dataOptions}
        chartData={scattermapData(dataOptions, dataSet)}
        designOptions={designOptions}
      />,
    );

    await waitFor(() => {
      const markers = document.querySelectorAll('.leaflet-interactive');
      fireEvent.click(markers[0]);
      const tooltip = document.querySelector('.leaflet-tooltip');
      expect(tooltip?.innerHTML).toMatchSnapshot();
    });
  });
});
