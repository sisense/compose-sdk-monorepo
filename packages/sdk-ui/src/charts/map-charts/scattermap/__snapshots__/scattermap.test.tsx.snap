// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Scattermap Chart > render a scattermap and call tooltip 1`] = `
"<div style="minWidth: 100px; color: #5B6372; fontSize: 13px; lineHeight: 18px; margin: 4px 6px">

    <div style="text-align: left">
        <div><span style="color:currentColor; font-size: 15px; line-height: 18px">55.5859012851966, -105.750595856519</span></div>
        <div class="csdk-scattermap-tooltip-row">

    <span style="padding-right: 20px">
      <span style="color:#9EA2AB; font-size: 15px; line-height: 18px">Quantity</span>
    </span>

        <span style="color:currentColor; font-size: 15px; line-height: 18px">6.78K</span>
        </div>
        <div class="csdk-scattermap-tooltip-row">


        </div>
        <div class="csdk-scattermap-tooltip-row">



        </div>
        <div>

        </div>
    </div>

  </div>"
`;
